src/comgr.cpp

```diff
@@ -0,0 +1,490 @@
+        return count;
+    }
+    Data GetData(const amd_comgr_data_kind_t kind, const std::size_t index) const
+    {
+        amd_comgr_data_t d;
+        ECI_THROW(amd_comgr_action_data_get_data(handle, kind, index, &d), kind);
+        return {d};
+    }
+};
+
+class ActionInfo : ComgrOwner
+{
+    amd_comgr_action_info_t handle = {0};
+
+    public:
+    ActionInfo() { EC_THROW(amd_comgr_create_action_info(&handle)); }
+    ~ActionInfo() { EC(amd_comgr_destroy_action_info(handle)); }
+    auto operator()() const { return handle; }
+    void SetLanguage(const amd_comgr_language_t language) const
+    {
+        ECI_THROW(amd_comgr_action_info_set_language(handle, language), language);
+    }
+    void SetIsaName(const std::string& isa) const
+    {
+        ECI_THROW_MSG(amd_comgr_action_info_set_isa_name(handle, isa.c_str()), isa, isa);
+    }
+    void SetLogging(const bool state) const
+    {
+        ECI_THROW(amd_comgr_action_info_set_logging(handle, state), state);
+    }
+    void SetOptionList(const std::vector<std::string>& options) const
+    {
+        std::vector<const char*> vp;
+        vp.reserve(options.size());
+        for(auto& opt : options) // cppcheck-suppress useStlAlgorithm
+            vp.push_back(opt.c_str());
+        LogOptions(vp.data(), vp.size());
+        ECI_THROW(amd_comgr_action_info_set_option_list(handle, vp.data(), vp.size()), vp.size());
+    }
+    void Do(const amd_comgr_action_kind_t kind, const Dataset& in, const Dataset& out) const
+    {
+        ECI_THROW_MSG(amd_comgr_do_action(kind, handle, in(), out()), kind, GetLog(out, true));
+        const auto log = GetLog(out);
+        if(!log.empty())
+            MIOPEN_LOG_I(to_string(kind) << ": " << GetLog(out));
+    }
+};
+
+/// If called from the context of comgr error handling:
+/// - We do not allow the comgr-induced exceptions to escape.
+/// - If obtaining log fails, write some diagnostics into output.
+static std::string GetLog(const Dataset& dataset, const bool comgr_error_handling)
+{
+    std::string text;
+    try
+    {
+        /// Assumption: the log is the the first in the dataset.
+        /// This is not specified in comgr API.
+        /// Let's follow the KISS principle for now; it works.
+        ///
+        /// \todo Clarify API and update implementation.
+        const auto count = dataset.GetDataCount(AMD_COMGR_DATA_KIND_LOG);
+        if(count < 1)
+            return {comgr_error_handling ? "comgr warning: error log not found" : ""};
+
+        const auto data = dataset.GetData(AMD_COMGR_DATA_KIND_LOG, 0);
+        text            = data.GetString();
+        if(text.empty())
+            return {comgr_error_handling ? "comgr info: error log empty" : ""};
+    }
+    catch(ComgrError&)
+    {
+        if(comgr_error_handling)
+            return {"comgr error: failed to get error log"};
+        // deepcode ignore EmptyThrowOutsideCatch: false positive
+        throw;
```

Comment:
What the point in catching the error if you are just going to rethrow it?

Reply:
In the context of handling of build error, the function is invoked to get build log from comgr. If it fails, it doesn't rethrow because this would overwrite the original comgr status.

The use case is when some build error happens (e.g. linking error) but we unable to obtain log data from the dataset due to comgr error. We keep original error information (albeit only status code) The user will see error message with original status code plus `comgr error: failed to get error log`.

Rethrowing happens when/if this function is invoked during normal flow, i.e. when there is no build errors. In such a case, the catch block does nothing and allows all exceptions to escape. This means that comgr errors would effectively stop the build.

(I feel like it's worth to add non-trivial explanations into C++ comments, what do you think?)

Reply:
Explanatory comment added to the source code.